crates/bevy_ecs/src/schedule/fsm.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{
+    self as bevy_ecs,
+    change_detection::Mut,
+    schedule::{run_systems, SystemLabel},
+    system::Res,
+    world::World,
+};
+
+use std::{fmt::Debug, hash::Hash};
+use thiserror::Error;
+
+/// Types that can define states in a finite-state machine.
+pub trait State: 'static + Send + Sync + Clone + PartialEq + Eq + Debug + Hash {}
+impl<T> State for T where T: 'static + Send + Sync + Clone + PartialEq + Eq + Debug + Hash {}
+
+/// A [`SystemLabel`] for a state's "on enter" transition.
```

Comment:
I love this design. Can we have an equivalent label for the on_update equivalents?

Reply:
States no longer have a "position" in the schedule, so there is no equivalent `on_update` set. 

To have something only run if a state is active, you'd attach the `state_equals(my_state_variant)` as a condition (or your own).

Reply:
Ah, makes sense.

My initial thought was that you could generate a label for those systems, but I don't really think it's worth it.